NAV: allow navigation from RsAbstractable to its implementations

Kobzol opened the pull request:
This is a revival of an old PR by @kumbayo (https://github.com/intellij-rust/intellij-rust/pull/2163), which allows navigation from functions/constants/associated types to their corresponding implementations.

The corresponding implementation place is rendered as `<Trait for Type>::function`.

Closes: https://github.com/intellij-rust/intellij-rust/pull/2163
Fixes: https://github.com/intellij-rust/intellij-rust/issues/2122
Part of: https://github.com/intellij-rust/intellij-rust/issues/2849

vlad20012 commented on src/main/kotlin/org/rust/ide/navigation/goto/RsGotoTargetRendererProvider.kt:
```diff
@@ -20,8 +21,11 @@ class RsGotoTargetRendererProvider : GotoTargetRendererProvider {
         }
-        if (element is RsImplItem) return ImplRenderer()
-        return null
+        return when (element) {
+            is RsImplItem,
+            is RsAbstractable -> NamedElementRenderer()
+            else -> null
+        }
```
This change causes ClassCastException when typing `Ctrl + Alt + B` on a trait.

Also, I don't understand why this is needed because we already have fine rendering for 
`Ctrl + Alt + B` implementation variants
![image](https://user-images.githubusercontent.com/3221931/86228759-71af6f80-bb97-11ea-8df9-c3ed0ccbd89e.png)

Kobzol replied:
You're right, this doesn't seem to be needed.

vlad20012 commented on src/main/kotlin/org/rust/ide/presentation/Utils.kt:
```diff
@@ -66,8 +66,27 @@ fun getPresentationForStructure(psi: RsElement): ItemPresentation {
+// Short presentation for display in the struct viewer
 private fun presentableName(psi: RsElement): String? {
     return when (psi) {
+        is RsAbstractable -> {
+            val o = psi.owner
+            if (o is RsAbstractableOwner.Impl) {
+                val typeName = o.impl.typeReference?.text ?: return null
+                val traitName = o.impl.traitRef?.text
+                buildString {
+                    append("<")
+                    if (traitName != null) {
+                        append("$traitName for ")
+                    }
+                    append(typeName)
+                    append(">::${psi.name}")
+                    append(typeParameterBounds(o.impl))
+                }
+            } else {
+                psi.name
+            }
+        }
```
This change causes test failures. And I don't think it's reasonable to change structure view rendering. It's possible to change renderer of implementation variants with `NavigationGutterIconBuilder.setCellRenderer` (in `RsImplsLineMarkerProvider.collectSlowLineMarkers`, see `info` variable)

Kobzol replied:
Thanks, I changed it so that it displays the presentation of the abstractable's owner.

Kobzol commented:
When you use `Ctrl + Alt + B`, it says `Choose implementations of <abstractable/trait name>`. When you click on Go to, it only shows `Go to implementation`, so when I have a trait like this:

```rust
trait T {              // X0
    fn foo(&self);     // X1
}
```
If you click on X0/X1, it will now show the same set of targets and it might now be obvious on what have you clicked. Maybe we could change the title to says `Go to implementation of T`/`Go to implementation of foo`?

vlad20012 commented on src/main/kotlin/org/rust/ide/lineMarkers/RsImplsLineMarkerProvider.kt:
```diff
@@ -42,21 +42,44 @@ class RsImplsLineMarkerProvider : LineMarkerProvider {
 
     companion object {
-        fun implsQuery(psi: PsiElement): Query<RsImplItem>? {
+        private class GoToImplRenderer: DefaultPsiElementCellRenderer() {
```
Can be moved outside of the companion object =D
(just to RsImplsLineMarkerProvider nested class)

Kobzol commented:
Fixed tests and added `Go to implementation of <element.text>`. Since the marker is provided for leaf elements (identifiers), I used the text of the identifier as the name.

Kobzol commented on src/test/kotlin/org/rust/ide/lineMarkers/RsMethodLineSeparatorProviderTest.kt:
```diff
@@ -27,14 +27,14 @@ class RsMethodLineSeparatorProviderTest : RsLineMarkerProviderTestBase() {
+                const C1: i32 = 1; // - Has implementations
+                fn f1(); // - null,Has implementations
+                fn f2(); // - null,Has implementations
+                fn f3() { // - Has implementations
```
I'm not really sure what is happening here and why the null has disappeared. I wonder if the comments can somehow affect the behaviour of `RsMethodLineSeparatorProvider`? Maybe we should strip the `// -` comments in the test?

vlad20012 replied:
Hmm, maybe we should strip them. But also, and more reasonable, we should disable other line marker providers in these tests. Let's fix it later

vlad20012 commented:
bors r+

Kobzol commented:
bors retry

Kobzol commented:
Yay, my first "resurrection" PR merged 🎉
